### DISCUSSION TITLE
Any way to write a hierarchical browser for Deno KV?

### DISCUSSION THREAD

**Discussion by skybrian** [denoland/deno#21729] [Q&A]
Deno KV's keys are lists, but the way they're commonly used suggests a directory tree. For example, the key for a user might be ["users", 123], which looks sort of like there are users in a users directory.

Is there some way to do the equivalent of "ls /" in a Unix filesystem? It doesn't look like the list operation could be used to do this, so it's not clear how one would write a browser for an arbitrary KV database, where you don't already know which keys are in it.

**Comment by nhrones** [comment]
See this example of a KV-Tree viewer
https://nhrones.github.io/KvRPC_TreeClient/

On the top-right, see the link for  **_see the code_**,  it will take you to the repo.

**Reply by skybrian** [reply]
That's an interesting example. It looks like it reads all the keys in the database into memory, though? It seems like this wouldn't scale, though it's good enough for small databases.

https://github.com/nhrones/BuenoRPC/blob/main/kvProcedures.ts#L71

```
/**  bulk fetch */
export async function getAllKeys() {
   const allKeys = []
   if (!db) await initDB()
   const entries = db.list({ prefix: [] })
   for await (const entry of entries) {
      allKeys.push(entry.key)
   }
   return allKeys
}
```

Deno KV is apparently missing an operation to do a directory listing in a scalable way.

**Comment by nhrones** [comment]
The RPC API could very easily be made to fetch by hierarchy tiers.  First tier (key[0]) would populate tree.  Selecting any tree-node would fetch all children of that prefix.

**Reply by skybrian** [reply]
Yes, that would help to make the RPC API that you wrote more efficient. However, it would still be doing the equivalent of a full table scan at the database level, and I don't see an easy way to change that without improvements to Deno KV itself.
